traits/trait_list_object.py

```diff
@@ -53,460 +88,750 @@ def __init__(self, index=0, removed=None, added=None):
         else:
-            self.len_error(len(self) - 1)
+            new_len = len(trait_list)
+            if removed is Undefined:
+                new_len += 1
+            if value is Undefined:
+                new_len -= 1
+        if not trait.minlen <= new_len <= trait.maxlen:
+            raise TraitError(
+                "The '%s' trait of %s instance must be %s, "
+                "but you attempted to change its length to %d element%s."
+                % (
+                    self.name,
+                    class_of(object),
+                    self.trait.full_info(object, self.name, Undefined),
+                    new_len,
+                    "s"[new_len == 1:],
+                )
+            )
 
-    def rename(self, name):
-        trait = self.object()._trait(name, 0)
-        if trait is not None:
-            self.name = name
-            self.trait = trait.handler
+        # validate the new value(s)
+        validate = trait.item_trait.handler.validate
+        if validate is None:
+            return value
 
-    def clear(self):
-        if self.trait.minlen > 0:
-            self.len_error(0)
+        try:
+            if isinstance(index, slice):
+                return [
+                    validate(object, self.name, item) for item in value
+                ]
+            elif value is Undefined:
+                return Undefined
+            else:
+                return validate(object, self.name, value)
+        except TraitError as excp:
+            excp.set_prefix("Each element of the")
+            raise excp
 
-        if len(self) == 0:
-            list.clear(self)
+    def notifier(self, trait_list, index, removed, added):
+        """ Converts and consolidates the parameters to a TraitListEvent and
+        then fires the event.
+
+        Parameters
+        ----------
+        trait_list : list
+            The list
+        index : int or slice
+            Index or slice that was modified
+        removed : value or list of values
+            Value or list of values that were removed
+        added : value or list of values
+            Value or list of values that were added
+
+        """
+        is_trait_none = self.trait is None
+        is_name_items_none = self.name_items is None
+        if not hasattr(self, "trait") or is_trait_none or is_name_items_none:
             return
 
-        removed = self.copy()
-        list.clear(self)
+        object = self.object()
+        if object is None:
+            return
 
-        if self.name_items is not None:
-            self._send_trait_items_event(
-                self.name_items, TraitListEvent(0, removed)
-            )
+        # bug-for-bug conversion of parameters to TraitListEvent
+        if isinstance(index, slice):
+            if index.step in {1, None}:
+                index = min(index.start, index.stop)
+            else:
+                if added and not isinstance(added, list):
+                    added = [added]
```

Comment:
Should `added` always be a list, the bug-for-bug conversion for `List(List)` handling would have to be implemented differently. But this should be possible, because we will have more information here.

Reply:
Yes, modified this.